### PR TITLE
Port all_to_all_single to native c10d_functional

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113057
* #113215
* #113735
* __->__ #113438

Summary:
- Ported `all_to_all_single` to native c10d_functional
- Added Inductor support for the native `all_to_all_single` via the new collective IR's `create_out_of_place()`
- Since the new collective IR derives from `FallbackKernel` which implements a generic `free_unbacked_symbols`, no additional unbacked symbol handling for all_to_all_single is required

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by yf225 — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
This op doesn't actually have dynamic shapes, it looks in order, I only skimmed

File: test/distributed/test_c10d_functional_native.py [hunk 1]
```diff
@@ -249,6 +249,33 @@ def func(args: List[torch.Tensor]) -> torch.Tensor:
             .run(code)
         )
 
+    @skip_if_lt_x_gpu(2)
+    def test_all_to_all_single(self) -> None:
+        self._init_process_group()
+        torch.cuda.set_device(self.device)
+
```
Comment (XilunWu) [inline test/distributed/test_c10d_functional_native.py:256]: curious if functional collective is supported as well on CPU backend `gloo`.

**Review by XilunWu — COMMENTED** [review]
Nice work Yifu! Have a noob question on func coll.

File: test/distributed/test_c10d_functional_native.py (hunk 1 above)
Comment (yifuwang) [reply test/distributed/test_c10d_functional_native.py:256]: Yes it's backend-agnostic. But yeah we should add some gloo tests.

**Pull request closed without merging**